File: spine-libgdx/spine-libgdx/src/com/esotericsoftware/spine/attachments/ClippingAttachment.java
/******************************************************************************
 * Spine Runtimes License Agreement
 * Last updated September 24, 2021. Replaces all prior versions.
 *
 * Copyright (c) 2013-2021, Esoteric Software LLC
 *
 * Integration of the Spine Runtimes into software or otherwise creating
 * derivative works of the Spine Runtimes is permitted under the terms and
 * conditions of Section 2 of the Spine Editor License Agreement:
 * http://esotericsoftware.com/spine-editor-license
 *
 * Otherwise, it is permitted to integrate the Spine Runtimes into software
 * or otherwise create derivative works of the Spine Runtimes (collectively,
 * "Products"), provided that each user of the Products must obtain their own
 * Spine Editor license and redistribution of the Products in any form must
 * include this license and copyright notice.
 *
 * THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
 * BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

package com.esotericsoftware.spine.attachments;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.utils.Null;

import com.esotericsoftware.spine.SlotData;

/** An attachment with vertices that make up a polygon used for clipping the rendering of other attachments. */
public class ClippingAttachment extends VertexAttachment {
	@Null SlotData endSlot;

	// Nonessential.
	final Color color = new Color(0.2275f, 0.2275f, 0.8078f, 1); // ce3a3aff

	public ClippingAttachment (String name) {
		super(name);
	}

	/** Copy constructor. */
	protected ClippingAttachment (ClippingAttachment other) {
		super(other);
		endSlot = other.endSlot;
		color.set(other.color);
	}

	/** Clipping is performed between the clipping attachment's slot and the end slot. If null clipping is done until the end of
	 * the skeleton's rendering. */
	public @Null SlotData getEndSlot () {
		return endSlot;
	}

	public void setEndSlot (@Null SlotData endSlot) {
		this.endSlot = endSlot;
	}

	/** The color of the clipping attachment as it was in Spine, or a default color if nonessential data was not exported. Clipping
	 * attachments are not usually rendered at runtime. */
	public Color getColor () {
		return color;
	}

	public ClippingAttachment copy () {
		return new ClippingAttachment(this);
	}
}

File: spine-csharp/src/Attachments/ClippingAttachment.cs
/******************************************************************************
 * Spine Runtimes License Agreement
 * Last updated April 5, 2025. Replaces all prior versions.
 *
 * Copyright (c) 2013-2026, Esoteric Software LLC
 *
 * Integration of the Spine Runtimes into software or otherwise creating
 * derivative works of the Spine Runtimes is permitted under the terms and
 * conditions of Section 2 of the Spine Editor License Agreement:
 * http://esotericsoftware.com/spine-editor-license
 *
 * Otherwise, it is permitted to integrate the Spine Runtimes into software
 * or otherwise create derivative works of the Spine Runtimes (collectively,
 * "Products"), provided that each user of the Products must obtain their own
 * Spine Editor license and redistribution of the Products in any form must
 * include this license and copyright notice.
 *
 * THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
 * BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

using System;

namespace Spine {
	public class ClippingAttachment : VertexAttachment {
		internal SlotData endSlot;
		internal bool convex, inverse;

		/// <summary>Clipping is performed between the clipping attachment's slot and the end slot. If null, clipping is done until the end of
		/// the skeleton's rendering.</summary>
		public SlotData EndSlot { get { return endSlot; } set { endSlot = value; } }

		/// <summary>
		/// When true the clipping polygon is treated as convex for more efficient clipping. If the polygon deforms to concave then the
		/// convex hull is used.When false the clipping polygon can be concave and if so has an additional CPU cost.Inverse clipping
		/// always uses convex.
		/// </summary>
		public bool Convex { get { return convex; } set { convex = value; } }

		/// <summary>
		/// When false, everything inside the clipping polygon is visible. When true, everything outside the clipping polygon is
		/// visible and clipping is convex.
		/// </summary>
		public bool Inverse { get { return inverse; } set { inverse = value; } }

		public ClippingAttachment (string name) : base(name) {
		}

		/// <summary>Copy constructor.</summary>
		protected ClippingAttachment (ClippingAttachment other)
			: base(other) {
			endSlot = other.endSlot;
			convex = other.convex;
			inverse = other.inverse;
		}

		public override Attachment Copy () {
			return new ClippingAttachment(this);
		}
	}
}
